File: backend/data/challenges.py
"""
Challenge data store for Battle-Front.

Each challenge has:
  - id, title, difficulty
  - description (shown to the user)
  - evaluation_criteria (list of strings fed to the Gemini prompt)
  - difficulty_scale: a function (score) -> (adjusted_score, was_adjusted)
"""

CHALLENGES: dict[str, dict] = {
    # ── 1. Interactive Counter (Easy) ───────────────────────────────────────
    "counter-button": {
        "id": "counter-button",
        "title": "Interactive Counter",
        "difficulty": "Easy",
        "description": (
            "Build a React counter component using useState. "
            "The button displays current count and increments by 1 on each click."
        ),
        "evaluation_criteria": [
            "useState hook used to manage count state",
            'Button displays "Current Count: {count}" format',
            "Count initializes at 0",
            "onClick handler correctly increments by exactly 1",
            "Functional component (not class-based)",
            "No external state libraries used",
        ],
    },
    # ── 2. Dynamic Todo List (Medium) ───────────────────────────────────────
    "todo-list": {
        "id": "todo-list",
        "title": "Dynamic Todo List",
        "difficulty": "Medium",
        "description": (
            "Build a todo list where users can add items via an input + button "
            "and remove them individually. State managed with React hooks."
        ),
        "evaluation_criteria": [
            "useState for managing the list array",
            "Add item appends to list without mutating state directly",
            "Remove item filters by index or id without mutation",
            "Input clears after adding an item",
            "Empty input submissions are prevented",
            "Each item has a visible remove/delete button",
            "List renders dynamically from state",
        ],
    },
    # ── 3. CSS Layout Ninja (Hard) ──────────────────────────────────────────
    "css-layout": {
        "id": "css-layout",
        "title": "CSS Layout Ninja",
        "difficulty": "Hard",
        "description": (
            "Replicate a complex dashboard layout using pure CSS Grid. "
            "Must include: header spanning full width, sidebar, main content "
            "area, and footer. No flexbox for the outer layout. No CSS frameworks."
        ),
        "evaluation_criteria": [
            "CSS Grid used for outer layout (not flexbox)",
            "Header spans full column width",
            "Sidebar occupies left column, main content takes remaining space",
            "Footer spans full width at bottom",
            "Grid template areas or explicit track sizing used correctly",
            "No external CSS frameworks or libraries",
            "Layout holds at reasonable viewport widths",
        ],
    },
    # ── 4. Debounce Function (Medium) ───────────────────────────────────────
    "debounce": {
        "id": "debounce",
        "title": "Debounce Function",
        "difficulty": "Medium",
        "description": (
            "Implement debounce(fn, delay) — returns a function that delays "
            "invoking fn until after delay ms have passed since the last call."
        ),
        "evaluation_criteria": [
            "Returns a new function (higher-order function)",
            "Delay resets on every new call before timeout fires",
            "Uses closure to persist the timer reference",
            "clearTimeout called before each new setTimeout",
            "Handles edge cases: rapid calls, zero delay",
            "No memory leaks from lingering timers",
            "Arguments correctly forwarded to original fn",
        ],
    },
    # ── 5. Infinite Scroll (Medium) ─────────────────────────────────────────
    "infinite-scroll": {
        "id": "infinite-scroll",
        "title": "Infinite Scroll",
        "difficulty": "Medium",
        "description": (
            "Build a scrollable list that loads more items automatically as "
            "the user scrolls to the bottom. Must use IntersectionObserver "
            "API, not scroll event listeners."
        ),
        "evaluation_criteria": [
            "IntersectionObserver used (not scroll events)",
            "Sentinel element at the bottom triggers loading",
            "Loading flag prevents duplicate concurrent fetches",
            "New items appended to existing list (not replaced)",
            "Async fetch simulated with setTimeout or Promise",
            "Observer disconnected or guarded against memory leaks",
            "End-of-data state handled gracefully",
        ],
    },
    # ── 6. Virtualized List (Hard) ──────────────────────────────────────────
    "virtualized-list": {
        "id": "virtualized-list",
        "title": "Virtualized List",
        "difficulty": "Hard",
        "description": (
            "Render a performant scrollable list of 100,000 items using "
            "windowing. Only visible DOM nodes should exist at any time. "
            "Each item is 40px tall. No external libraries."
        ),
        "evaluation_criteria": [
            "Only visible items in DOM at any time",
            "Total container height equals 100000 * 40px",
            "Items absolutely positioned based on index",
            "Correct visible range calculation from scrollTop",
            "Buffer rows above and below viewport to prevent flicker",
            "Scroll listener used efficiently (no layout thrash)",
            "Works correctly after scrolling to any position rapidly",
        ],
    },
}


def get_challenge(challenge_id: str) -> dict | None:
    """Return challenge dict or None if not found."""
    return CHALLENGES.get(challenge_id)


def apply_difficulty_scaling(score: int, difficulty: str) -> tuple[int, bool]:
    """
    Apply difficulty-based scoring adjustments.
    Returns (adjusted_score, was_adjusted).
    """
    if difficulty == "Medium" and score >= 60:
        return min(score + 10, 100), True
    if difficulty == "Hard" and score >= 50:
        return min(score + 15, 100), True
    return score, False


def derive_verdict(score: int) -> str:
    """Derive verdict string from numerical score."""
    if score >= 80:
        return "Pass"
    if score >= 40:
        return "Partial"
    return "Fail"

File: frontend/lib/challenges.js
const challenges = [
  {
    id: "counter-button",
    title: "Interactive Counter",
    difficulty: "Easy",
    shortDescription: "Build a standard interactive counter using React hooks.",
    description: `### Task: Simple React Counter\n\nCreate a functional counter button using React's useState hook. The button should display the current count and increment by 1 each time it is clicked.\n\n### Requirements:\n1. Initialize count at 0\n2. Update the button text to show: "Current Count: {count}"\n3. The count must increment by exactly 1 per click`,
    starterCode: `import React, { useState } from 'react';

/**
 * Task: Implement a counter that increments on click.
 * Use React.useState().
 */
export default function Counter() {
    // WRITE YOUR CODE HERE
    return (
        <button id="counter-btn" style={{ padding: '12px 24px', borderRadius: '8px', background: '#4F46E5', color: 'white', border: 'none', cursor: 'pointer' }}>
            Count: 0
        </button>
    );
}`,
    constraints: [
      "Must use React hooks",
      "Must be functional component",
      "Should not use external libraries for state",
    ],
    hints: [
      "Import { useState } from 'react'",
      "Call useState(0) to initialize state",
      "Add an onClick handler to the button",
    ],
  },
  {
    id: "todo-list",
    title: "Dynamic Todo List",
    difficulty: "Medium",
    shortDescription:
      "Implement a list where users can add and remove items dynamically.",
    description: `### Task: Dynamic Todo List\n\nBuild a basic Todo list application where users can input text into a field and add it to a list. Each item should have a "Delete" button that removes it from the list.\n\n### Requirements:\n1. Controlled input field for new tasks\n2. "Add" button to append task to list\n3. Map over tasks to render them\n4. Remove individual tasks via a delete button`,
    starterCode: `import React, { useState } from 'react';

export default function TodoList() {
    const [tasks, setTasks] = useState([]);
    const [inputValue, setInputValue] = useState('');

    const addTask = () => {
        if (!inputValue) return;
        setTasks([...tasks, inputValue]);
        setInputValue('');
    };

    const removeTask = (index) => {
        setTasks(tasks.filter((_, idx) => idx !== index));
    };

    return (
        <div style={{ padding: '20px' }}>
            <input 
                value={inputValue} 
                onChange={(e) => setInputValue(e.target.value)}
                style={{ background: '#333', color: '#fff', border: '1px solid #555', padding: '8px' }}
            />
            <button onClick={addTask} style={{ marginLeft: '10px' }}>Add</button>
            <ul>
                {tasks.map((task, i) => (
                    <li key={i}>
                        {task} <button onClick={() => removeTask(i)}>X</button>
                    </li>
                ))}
            </ul>
        </div>
    );
}`,
    constraints: [
      "Items must be unique-ish (index is fine for basic)",
      "State must be handled in React",
    ],
    hints: [
      "Use an array for state",
      "Use .filter() for removal",
      "Prevent adding empty strings",
    ],
  },
  {
    id: "css-layout-ninja",
    title: "CSS Layout Ninja",
    difficulty: "Hard",
    shortDescription:
      "Replicate a complex grid-based dashboard layout using pure CSS.",
    description: `### Task: Grid-Based Dashboard\n\nYour task is to create a responsive dashboard layout using CSS Grid and Flexbox. \n\n### Requirements:\n1. A sidebar that takes 25% width on desktop and hides on mobile.\n2. A main content area with at least 4 cards.\n3. A header that sticks to the top.\n4. Use Tailwind CSS classes if possible, or inline styles for this sandbox.`,
    starterCode: `import React from 'react';

export default function Dashboard() {
    return (
        <div style={{ display: 'grid', gridTemplateColumns: '250px 1fr', height: '100vh', background: '#0f172a' }}>
            <aside style={{ background: '#1e293b', borderRight: '1px solid #334155', padding: '16px' }}>
                <h3 style={{ color: '#94a3b8' }}>Sidebar</h3>
            </aside>
            <main style={{ padding: '24px' }}>
                <header style={{ marginBottom: '24px', display: 'flex', justifyContent: 'space-between' }}>
                    <h1 style={{ color: '#fff' }}>Overview</h1>
                    <div style={{ display: 'flex', gap: '8px' }}>
                        <div style={{ width: '32px', height: '32px', borderRadius: '50%', background: '#4F46E5' }}></div>
                    </div>
                </header>
                <div style={{ display: 'grid', gridTemplateColumns: 'repeat(auto-fill, minmax(200px, 1fr))', gap: '16px' }}>
                   {/* Add your cards here */}
                   <div style={{ background: '#1e293b', padding: '20px', borderRadius: '12px', border: '1px solid #334155', color: '#fff' }}>
                        Card 1
                   </div>
                </div>
            </main>
        </div>
    );
}`,
    constraints: ["Must be responsive", "No external grid libraries"],
    hints: [
      "Use display: grid",
      "Use media queries via @media or responsive utilities",
    ],
  },
];

export default challenges;
